File: proj.android_studio/ee-x-google-analytics/src/main/java/com/ee/google/analytics/GoogleAnalyticsTracker.java
package com.ee.google.analytics;

import androidx.annotation.NonNull;

import com.ee.core.IMessageBridge;
import com.ee.core.MessageBridge;
import com.ee.core.MessageHandler;
import com.ee.core.internal.JsonUtils;
import com.ee.core.internal.Utils;
import com.google.android.gms.analytics.Tracker;

import java.util.HashMap;
import java.util.Map;

import static com.google.common.truth.Truth.assertThat;

/**
 * Created by Zinge on 10/25/17.
 */

public class GoogleAnalyticsTracker {
    private static final String k__key = "key";
    private static final String k__value = "value";

    private IMessageBridge _bridge;
    private String _trackingId;
    private Tracker _tracker;

    GoogleAnalyticsTracker(@NonNull IMessageBridge bridge,
                           @NonNull com.google.android.gms.analytics.GoogleAnalytics analytics,
                           @NonNull String trackingId) {
        _bridge = bridge;
        _trackingId = trackingId;
        _tracker = analytics.newTracker(trackingId);
        registerHandlers();
    }

    void destroy() {
        deregisterHandlers();
        _bridge = null;
        _tracker = null;
        _trackingId = null;
    }

    @NonNull
    private String k__setParameter() {
        return "GoogleAnalytics_setParameter_" + _trackingId;
    }

    @NonNull
    private String k__setAllowIDFACollection() {
        return "GoogleAnalytics_setAllowIDFACollection_" + _trackingId;
    }

    @NonNull
    private String k__send() {
        return "GoogleAnalytics_send_" + _trackingId;
    }

    private void registerHandlers() {
        _bridge.registerHandler(message -> {
            Map<String, Object> dict = JsonUtils.convertStringToDictionary(message);
            assertThat(dict).isNotNull();
            String key = (String) dict.get(k__key);
            String value = (String) dict.get(k__value);
            setParameter(key, value);
            return "";
        }, k__setParameter());

        _bridge.registerHandler(message -> {
            setAdvertisingIdCollectionEnabled(Utils.toBoolean(message));
            return "";
        }, k__setAllowIDFACollection());

        _bridge.registerHandler(message -> {
            Map<String, Object> rawDict = JsonUtils.convertStringToDictionary(message);
            assertThat(rawDict).isNotNull();

            Map<String, String> dict = new HashMap<>();
            for (String key : rawDict.keySet()) {
                dict.put(key, (String) rawDict.get(key));
            }
            send(dict);
            return "";
        }, k__send());
    }

    private void deregisterHandlers() {
        _bridge.deregisterHandler(k__setParameter());
        _bridge.deregisterHandler(k__setAllowIDFACollection());
        _bridge.deregisterHandler(k__send());
    }

    @SuppressWarnings("WeakerAccess")
    public void setParameter(@NonNull String key, @NonNull String value) {
        _tracker.set(key, value);
    }

    @SuppressWarnings("WeakerAccess")
    public void setAdvertisingIdCollectionEnabled(boolean enabled) {
        _tracker.enableAdvertisingIdCollection(enabled);
    }

    public void setExceptionReportingEnabled(boolean enabled) {
        _tracker.enableExceptionReporting(enabled);
    }

    @SuppressWarnings("WeakerAccess")
    public void send(@NonNull Map<String, String> dict) {
        _tracker.send(dict);
    }
}

File: proj.android_studio/ee-x-google-analytics/src/main/java/com/ee/internal/GoogleAnalyticsTracker.kt
package com.ee.internal

import androidx.annotation.AnyThread
import com.ee.IMessageBridge
import com.ee.Thread
import com.ee.Utils
import com.ee.registerHandler
import com.google.android.gms.analytics.GoogleAnalytics
import kotlinx.serialization.ImplicitReflectionSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.UnstableDefault

/**
 * Created by Zinge on 10/25/17.
 */
@ImplicitReflectionSerializer
@UnstableDefault
class GoogleAnalyticsTracker(
    private val _bridge: IMessageBridge,
    private val _analytics: GoogleAnalytics,
    private val _trackingId: String) {
    private val _tracker = _analytics.newTracker(_trackingId)

    init {
        registerHandlers()
    }

    @AnyThread
    fun destroy() {
        deregisterHandlers()
    }

    private val k__setParameter: String
        @AnyThread get() = "GoogleAnalytics_setParameter_$_trackingId"

    private val k__setAllowIDFACollection: String
        @AnyThread get() = "GoogleAnalytics_setAllowIDFACollection_$_trackingId"

    private val k__send: String
        @AnyThread get() = "GoogleAnalytics_send_$_trackingId"

    @AnyThread
    private fun registerHandlers() {
        _bridge.registerHandler(k__setParameter) { message ->
            @Serializable
            class Request(
                val key: String,
                val value: String
            )

            val request = deserialize<Request>(message)
            setParameter(request.key, request.value)
            ""
        }
        _bridge.registerHandler(k__setAllowIDFACollection) { message ->
            setAdvertisingIdCollectionEnabled(Utils.toBoolean(message))
            ""
        }
        _bridge.registerHandler(k__send) { message ->
            val params = deserializeMap<String, String>(message)
            send(params)
            ""
        }
    }

    @AnyThread
    private fun deregisterHandlers() {
        _bridge.deregisterHandler(k__setParameter)
        _bridge.deregisterHandler(k__setAllowIDFACollection)
        _bridge.deregisterHandler(k__send)
    }

    @AnyThread
    fun setParameter(key: String, value: String) {
        Thread.runOnMainThread(Runnable {
            _tracker[key] = value
        })
    }

    @AnyThread
    private fun setAdvertisingIdCollectionEnabled(enabled: Boolean) {
        Thread.runOnMainThread(Runnable {
            _tracker.enableAdvertisingIdCollection(enabled)
        })
    }

    @AnyThread
    fun setExceptionReportingEnabled(enabled: Boolean) {
        Thread.runOnMainThread(Runnable {
            _tracker.enableExceptionReporting(enabled)
        })
    }

    @AnyThread
    fun send(params: Map<String, String>) {
        Thread.runOnMainThread(Runnable {
            _tracker.send(params)
        })
    }
}